pipeliner.go

```diff
@@ -0,0 +1,241 @@
+package radix
+
+import (
+	"strings"
+	"sync"
+	"time"
+)
+
+var blockingCmds = map[string]bool{
+	"WAIT": true,
+
+	// taken from https://github.com/joomcode/redispipe#limitations
+	"BLPOP":      true,
+	"BRPOP":      true,
+	"BRPOPLPUSH": true,
+
+	"BZPOPMIN": true,
+	"BZPOPMAX": true,
+
+	"XREAD":      true,
+	"XREADGROUP": true,
+
+	"SAVE":  true,
+}
+
+type pipeliner struct {
+	c Client
+
+	limit  int
+	window time.Duration
+
+	// reqsBufCh contains buffers for collecting commands and acts as a semaphore
+	// to limit the number of concurrent flushes.
+	reqsBufCh chan []CmdAction
+
+	reqCh chan *pipelinerCmd
+	reqWG sync.WaitGroup
+
+	l      sync.RWMutex
+	closed bool
+}
+
+var _ Client = (*pipeliner)(nil)
+
+func newPipeliner(c Client, concurrency, limit int, window time.Duration) *pipeliner {
+	if concurrency < 1 {
+		concurrency = 1
+	}
+
+	p := &pipeliner{
+		c: c,
+
+		limit:  limit,
+		window: window,
+
+		reqsBufCh: make(chan []CmdAction, concurrency),
+
+		reqCh: make(chan *pipelinerCmd, 32), // https://xkcd.com/221/
+	}
+
+	p.reqWG.Add(1)
+	go func() {
+		defer p.reqWG.Done()
+		p.reqLoop()
+	}()
+
+	for i := 0; i < cap(p.reqsBufCh); i++ {
+		if p.limit > 0 {
+			p.reqsBufCh <- make([]CmdAction, 0, limit)
+		} else {
+			p.reqsBufCh <- nil
+		}
+	}
+
+	return p
+}
+
+// CanDo checks if the given Action can be executed / passed to p.Do.
+//
+// If CanDo returns false, the Action must not be given to Do.
+func (p *pipeliner) CanDo(a Action) bool {
```

Comment:
This method could be significantly simplified:

```go
// there is currently no way to get the command for CmdAction implementations
// from outside the radix package so we can not multiplex those commands. User
// defined peipelines are not pipelined to let the user better control them.
if cmdA, ok := a.(*cmdAction); ok {
     return !blockingCmds[strings.ToUpper(cmdA.cmd)]
}
return false

Reply:
Done